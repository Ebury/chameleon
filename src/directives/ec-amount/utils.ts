
import defaultOptions from './options';
import type { AmountDirectiveOptions } from './types';

function format(input: undefined, opts?: AmountDirectiveOptions): undefined;
function format(input: null, opts?: AmountDirectiveOptions): null;
function format(input: string | number, opts?: AmountDirectiveOptions): string;
function format(input: string | number | undefined | null, opts: AmountDirectiveOptions = defaultOptions): string | null | undefined {
  /* eslint-disable no-param-reassign */
  if (input === '' || input === null || input === undefined) {
    return input;
  }

  if (typeof input === 'number') {
    input = input.toString();
  }

  const negative = input.startsWith('-') ? '-' : '';

  input = sanitizeInput(input, opts.decimalSeparator);
  if (opts.precision > 0) {
    input = keepOnlyOneDecimalSeparator(input, opts.decimalSeparator);
  } else {
    input = removeDecimal(input, opts.decimalSeparator);
  }
  const parts = input.split(opts.decimalSeparator);
  let integralPart = parseToStr(parts[0]);
  let fractionalPart = parts[1];

  if (fractionalPart && fractionalPart.length > opts.precision) {
    fractionalPart = fractionalPart.substring(0, opts.precision);
  }

  integralPart = addGroupingSeparators(integralPart, opts.groupingSeparator);
  const result = negative + joinIntegralAndFractionalParts(integralPart, fractionalPart, opts.decimalSeparator);
  return result;
}

function parseToStr(number: string): string {
  const parsed = Number.parseFloat(number);
  if (Number.isNaN(parsed)) {
    return '';
  }
  return parsed.toString();
}

function keepOnlyOneDecimalSeparator(number: string, decimalSeparator: string): string {
  if (number.includes(decimalSeparator)) {
    const split = number.split(decimalSeparator);
    return split[0] + decimalSeparator + split.slice(1).join('');
  }
  return number;
}

function removeDecimal(number: string, decimalSeparator: string): string {
  const split = number.split(decimalSeparator);
  return split[0];
}

function unFormat(input: string, groupingSeparator: string, decimalSeparator: string): string {
  if (!input) {
    return input;
  }

  return input
    .replace(new RegExp(`\\${groupingSeparator}`, 'g'), '')
    .replace(new RegExp(`\\${decimalSeparator}`, 'g'), '.');
}

function sanitizeInput(input: string, decimalSeparator: string): string {
  return input.replace(new RegExp(`[^0-9${decimalSeparator}]`, 'g'), '') || '';
}

function addGroupingSeparators(integralPart: string, separator: string): string {
  return integralPart.replace(/(\d)(?=(?:\d{3})+\b)/gm, `$1${separator}`);
}

function joinIntegralAndFractionalParts(integralPart: string, fractionalPart: string, decimalSeparator: string): string {
  return fractionalPart !== undefined ? integralPart + decimalSeparator + fractionalPart : integralPart;
}

/* c8 ignore start */
function setCursor(el: HTMLInputElement, position: number) {
  const setSelectionRange = function setSelectionRange() {
    el.setSelectionRange(position, position);
  };

  if (el === document.activeElement) {
    setSelectionRange();
    setTimeout(setSelectionRange, 1); // Android Fix
  }
}
/* c8 ignore stop */

export {
  format,
  sanitizeInput,
  setCursor,
  unFormat,
};
