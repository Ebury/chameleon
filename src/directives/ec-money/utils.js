import defaults from './options';

function format(input, opt = defaults) {
  if (input === '' || input === null || input === undefined) {
    return input;
  }

  if (typeof input === 'number') {
    input = input.toString();
  }

  const negative = input.includes('-') ? '-' : '';

  input = sanitizeInput(input, opt.decimalSeparator);
  input = keepOnlyOneDecimalSeparator(input, opt.decimalSeparator);
  const parts = input.split(opt.decimalSeparator);
  let integer = parseToStr(parts[0]);
  let decimal = parts[1];

  if (decimal && decimal.length > opt.precision) {
    decimal = decimal.substr(0, opt.precision);
  }

  integer = addGroupingSeparators(integer, opt.groupingSeparator);
  const result = negative + joinIntegerAndDecimal(integer, decimal, opt.decimalSeparator);
  return result;
}

function parseToStr(number) {
  /* eslint no-restricted-globals: "off" */
  const parsed = parseFloat(number);
  if (isNaN(parsed)) {
    return '';
  }
  return parsed.toString();
}

function keepOnlyOneDecimalSeparator(number, decimalSeparator) {
  if (number.includes(decimalSeparator)) {
    const split = number.split(decimalSeparator);
    return split[0] + decimalSeparator + split.slice(1).join('');
  }
  return number;
}

function unFormat(input, groupingSeparator, decimalSeparator) {
  /* eslint no-param-reassign: "off" */
  input = (`${input}`).replace(new RegExp(`\\${groupingSeparator}`, 'g'), '');
  input = (`${input}`).replace(new RegExp(`\\${decimalSeparator}`, 'g'), '.');
  return input;
}

function sanitizeInput(input, decimalSeparator) {
  return input.replace(new RegExp(`[^0-9${decimalSeparator}]`, 'g'), '') || '';
}

function addGroupingSeparators(integer, separator) {
  return integer.replace(/(\d)(?=(?:\d{3})+\b)/gm, `$1${separator}`);
}

function joinIntegerAndDecimal(integer, decimal, decimalSeparator) {
  return decimal !== undefined ? integer + decimalSeparator + decimal : integer;
}

function setCursor(el, position) {
  const setSelectionRange = function setSelectionRange() {
    el.setSelectionRange(position, position);
  };

  if (el === document.activeElement) {
    setSelectionRange();
    setTimeout(setSelectionRange, 1); // Android Fix
  }
}

// https://developer.mozilla.org/en-US/docs/Web/Guide/Events/Creating_and_triggering_events#The_old-fashioned_way
function event(name) {
  const evt = document.createEvent('Event');
  evt.initEvent(name, true, true);
  return evt;
}

export {
  format,
  unFormat,
  setCursor,
  event,
  sanitizeInput,
};
