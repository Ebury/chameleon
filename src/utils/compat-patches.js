/* istanbul ignore file */

import {
  Dropdown as FvDropdown,
  TooltipDirective as FvTooltipDirective,
  VClosePopper,
  VTooltip,
} from 'floating-vue';

// we need to tell the @vue/compat that all components coming from floating-vue are Vue 3 components so it skips
// them when patching component instances. These components can't run in default mode for Vue 2, there is lots of
// errors coming from the console in that mode.
export default function applyCompatPatches() {
  const compatConfig = {
    MODE: 3,
    RENDER_FUNCTION: false,
  };

  VTooltip.compatConfig = compatConfig;
  VClosePopper.compatConfig = compatConfig;
  FvTooltipDirective.compatConfig = compatConfig;
  for (const component of Object.values(FvTooltipDirective.components)) {
    component.compatConfig = compatConfig;
  }

  FvDropdown.compatConfig = compatConfig;

  for (const component of Object.values(FvDropdown.components)) {
    component.compatConfig = compatConfig;
  }
}
