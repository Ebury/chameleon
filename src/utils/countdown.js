import mitt from 'mitt';

export default class Countdown {
  constructor() {
    const m = mitt();
    Object.assign(this, m);
  }

  reduceMinutesLeft(totalSeconds) {
    this.getminutes(totalSeconds);
    this.getseconds(totalSeconds, this.minutesLeft);
  }

  start(secondsToGo, showMinutes) {
    this.seconds = secondsToGo;
    this.secondsLeft = secondsToGo;
    this.currentTime = Math.ceil(Date.now() / 1000);
    this.startTime = Math.ceil(Date.now() / 1000);
    this.interval = setInterval(() => this.reduceSecondsLeft(showMinutes), 1000);
  }

  get timeDifference() {
    return this.currentTime - this.startTime;
  }

  reduceSecondsLeft(showMinutes) {
    this.currentTime = Math.ceil(Date.now() / 1000);
    this.totalSeconds = this.seconds - this.timeDifference;
    this.minutesLeft = null;
    this.secondsLeft = this.totalSeconds;

    if (showMinutes) {
      this.reduceMinutesLeft(this.totalSeconds);
    }

    this.emit('time-updated', {
      seconds: this.secondsLeft,
      minutes: this.minutesLeft,
      totalSeconds: this.totalSeconds,
    });

    if (this.totalSeconds <= 0) {
      this.stop();
      this.emit('time-expired');
    }
  }

  stop() {
    clearInterval(this.interval);
  }

  getminutes(totalSeconds) {
    this.minutesLeft = Math.floor(totalSeconds / 60);
  }

  getseconds(totalSeconds, minutes) {
    // take mins remaining (as seconds) away from total seconds remaining
    this.secondsLeft = totalSeconds - Math.round(minutes * 60);
  }
}
