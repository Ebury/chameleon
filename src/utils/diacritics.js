export function removeDiacritics(str) {
  if (typeof str === 'string' && str.normalize) {
    // https://stackoverflow.com/questions/990904/remove-accents-diacritics-in-a-string-in-javascript
    return str.normalize('NFD').replace(/[\u0300-\u036f]/g, '');
  }

  // IE11 users will not enjoy this feature:
  // https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/String/normalize
  return str;
}
