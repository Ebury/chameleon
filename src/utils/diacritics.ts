export function removeDiacritics(str: string): string {
  if (str && str.normalize) {
    // https://stackoverflow.com/questions/990904/remove-accents-diacritics-in-a-string-in-javascript
    return str.normalize('NFD').replace(/[\u0300-\u036f]/g, '');
  }

  // IE11 users will not enjoy this feature:
  // https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/String/normalize
  return str;
}
