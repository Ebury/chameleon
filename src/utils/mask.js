export function mask(text, maskSymbol = '*', visibleChars = 2) {
  if (!text) {
    throw new Error('Text is required');
  }
  if (!maskSymbol) {
    throw new Error('Mask symbol cannot be empty');
  }
  if (!visibleChars || visibleChars < 0) {
    throw new Error('Visible characters must be a number greater than zero');
  }

  const digitsToBeMasked = text.length - visibleChars;
  return maskSymbol.repeat(digitsToBeMasked) + text.substr(digitsToBeMasked);
}
