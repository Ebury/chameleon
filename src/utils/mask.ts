export function mask(text: string, maskSymbol: string = '*', visibleChars: number = 2): string {
  if (!text) {
    throw new Error('Text is required');
  }
  if (!maskSymbol) {
    throw new Error('Mask symbol cannot be empty');
  }
  if (!visibleChars || visibleChars < 0) {
    throw new Error('Visible characters must be a number greater than zero');
  }

  const digitsToBeMasked = text.length - visibleChars;
  return maskSymbol.repeat(digitsToBeMasked) + text.substring(digitsToBeMasked);
}
