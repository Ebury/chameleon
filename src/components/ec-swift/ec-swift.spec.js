import Vue from 'vue';
import { mount } from '@vue/test-utils';
import EcSwift from './ec-swift.vue';
import { withMockedConsole } from '../../../tests/utils/console';

describe('EcSwift', () => {
  function mountSwift(props, mountOpts) {
    return mount(EcSwift, {
      propsData: {
        title: 'Title example',
        type: 'info',
        ...props,
      },
      ...mountOpts,
    });
  }

  it('should throw if no props were given', () => {
    withMockedConsole((errorSpy) => {
      mount(EcSwift);
      expect(errorSpy).toHaveBeenCalledTimes(2);
      expect(errorSpy.mock.calls[0][0]).toContain('Missing required prop: "type"');
      expect(errorSpy.mock.calls[1][0]).toContain('Missing required prop: "title"');
    });
  });

  it('should display only with a title and the type given', () => {
    const wrapper = mountSwift({ title: 'Random Title', type: 'info' });
    expect(wrapper.element).toMatchSnapshot();
  });

  it('should display with a subtitle given', () => {
    const wrapper = mountSwift({ subtitle: 'Subtitle example' });
    expect(wrapper.element).toMatchSnapshot();
  });

  it('should display with a button with the text given in the buttonText', () => {
    const wrapper = mountSwift({ buttonText: 'Button text' });
    expect(wrapper.element).toMatchSnapshot();
  });

  it('should use the given type for the swift type and for the button too', () => {
    const wrapper = mountSwift({ type: 'error', buttonText: 'Warning button' });
    expect(wrapper.element).toMatchSnapshot();
  });

  it.each(['error', 'info', 'success', 'warning'])('should use the type "%s"', (type) => {
    const wrapper = mountSwift({ type });
    expect(wrapper.element).toMatchSnapshot();
  });

  it('should throw an error if type is not valid', () => {
    withMockedConsole((errorSpy) => {
      mountSwift({ type: 'invalid-value' });
      expect(errorSpy).toHaveBeenCalledTimes(1);
      expect(errorSpy.mock.calls[0][0]).toContain('Invalid prop: custom validator check failed for prop "type"');
    });
  });

  it('should render the dismiss icon when is given the prop dismissable', () => {
    const wrapper = mountSwift({ dismissable: true });
    expect(wrapper.element).toMatchSnapshot();
  });

  it('should dismiss the swift when user clicks on the dismiss icon ', () => {
    const wrapper = mount(Vue.extend({
      components: { EcSwift },
      data() {
        return { isOpen: true };
      },
      template: `
        <ec-swift v-model="isOpen" type="info" title="Custom random" dismissable />
      `,
    }));
    expect(wrapper.find('.ec-swift__dismiss-icon').exists()).toBe(true);
    wrapper.find('.ec-swift__dismiss-icon').trigger('click');
    expect(wrapper.isVisible()).toBe(false);
  });

  it('should emit the event when user clicks on the button', () => {
    const wrapper = mountSwift({ buttonText: 'Click here' });
    wrapper.find('.ec-swift__button').trigger('click');
    expect(wrapper.emitted('action').length).toBe(1);
  });

  it('should dismiss or show the swift when we change the v-model', () => {
    const wrapper = mount(Vue.extend({
      components: { EcSwift },
      data() {
        return { isOpen: true };
      },
      template: `
        <ec-swift v-model="isOpen" type="info" title="Custom random" dismissable />
      `,
    }));

    expect(wrapper.vm.isOpen).toBe(true);
    expect(wrapper.isVisible()).toBe(true);
    wrapper.find('.ec-swift__dismiss-icon').trigger('click');
    expect(wrapper.isVisible()).toBe(false);
    expect(wrapper.vm.isOpen).toBe(false);
  });
  it('should render with the slot given', () => {
    const wrapper = mountSwift({
      subtitle: 'Subtitle example',
    },
    {
      scopedSlots: {
        default: '<div slot-scope="{ title, subtitle }">Custom: {{ title }} - {{ subtitle }}</div>',
      },
    });

    expect(wrapper.element).toMatchSnapshot();
  });

  it('should render responsive by default', () => {
    const wrapper = mountSwift();
    expect(wrapper.classes('ec-swift--is-responsive')).toBe(true);
  });

  it('should not render responsive if the responsive prop is set to false', () => {
    const wrapper = mountSwift({ responsive: false });
    expect(wrapper.classes('ec-swift--is-responsive')).toBe(false);
  });
});
