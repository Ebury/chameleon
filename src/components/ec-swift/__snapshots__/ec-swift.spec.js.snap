// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EcSwift should display only with a title and the type given 1`] = `
<div
  class="ec-swift ec-swift--info ec-swift--is-responsive"
>
  <!---->
   
  <svg
    class="ec-swift__icon ec-icon"
  >
    <use
      xlink:href="#ec-simple-info"
    />
  </svg>
   
  <div
    class="ec-swift__content"
  >
    <div
      class="ec-swift__title"
    >
      Random Title
    </div>
     
    <!---->
  </div>
   
  <!---->
</div>
`;

exports[`EcSwift should display with a button with the text given in the buttonText 1`] = `
<div
  class="ec-swift ec-swift--info ec-swift--is-responsive"
>
  <!---->
   
  <svg
    class="ec-swift__icon ec-icon"
  >
    <use
      xlink:href="#ec-simple-info"
    />
  </svg>
   
  <div
    class="ec-swift__content"
  >
    <div
      class="ec-swift__title"
    >
      Title example
    </div>
     
    <!---->
  </div>
   
  <button
    class="ec-btn ec-btn--sm ec-btn--outline ec-btn--rounded ec-swift__button ec-btn--primary-reverse"
  >
    Button text
  </button>
</div>
`;

exports[`EcSwift should display with a subtitle given 1`] = `
<div
  class="ec-swift ec-swift--info ec-swift--is-responsive"
>
  <!---->
   
  <svg
    class="ec-swift__icon ec-icon ec-swift__icon--swift-has-subtitle"
  >
    <use
      xlink:href="#ec-simple-info"
    />
  </svg>
   
  <div
    class="ec-swift__content"
  >
    <div
      class="ec-swift__title"
    >
      Title example
    </div>
     
    <div
      class="ec-swift__subtitle"
    >
      Subtitle example
    </div>
  </div>
   
  <!---->
</div>
`;

exports[`EcSwift should render the dismiss icon when is given the prop dismissable 1`] = `
<div
  class="ec-swift ec-swift--info ec-swift--is-responsive"
>
  <a
    class="ec-swift__dismiss-icon ec-swift__dismiss-icon--info"
    href="#"
  >
    <svg
      class="ec-icon"
      height="16"
      width="16"
    >
      <use
        xlink:href="#ec-simple-close"
      />
    </svg>
  </a>
   
  <svg
    class="ec-swift__icon ec-icon"
  >
    <use
      xlink:href="#ec-simple-info"
    />
  </svg>
   
  <div
    class="ec-swift__content"
  >
    <div
      class="ec-swift__title"
    >
      Title example
    </div>
     
    <!---->
  </div>
   
  <!---->
</div>
`;

exports[`EcSwift should render with the slot given 1`] = `
<div
  class="ec-swift ec-swift--info ec-swift--is-responsive"
>
  <!---->
   
  <svg
    class="ec-swift__icon ec-icon ec-swift__icon--swift-has-subtitle"
  >
    <use
      xlink:href="#ec-simple-info"
    />
  </svg>
   
  <div
    class="ec-swift__content"
  >
    <div>
      Custom: Title example - Subtitle example
    </div>
  </div>
   
  <!---->
</div>
`;

exports[`EcSwift should use the given type for the swift type and for the button too 1`] = `
<div
  class="ec-swift ec-swift--error ec-swift--is-responsive"
>
  <!---->
   
  <svg
    class="ec-swift__icon ec-icon"
  >
    <use
      xlink:href="#ec-simple-error"
    />
  </svg>
   
  <div
    class="ec-swift__content"
  >
    <div
      class="ec-swift__title"
    >
      Title example
    </div>
     
    <!---->
  </div>
   
  <button
    class="ec-btn ec-btn--sm ec-btn--outline ec-btn--rounded ec-swift__button ec-btn--error-reverse"
  >
    Warning button
  </button>
</div>
`;

exports[`EcSwift should use the type "error" 1`] = `
<div
  class="ec-swift ec-swift--error ec-swift--is-responsive"
>
  <!---->
   
  <svg
    class="ec-swift__icon ec-icon"
  >
    <use
      xlink:href="#ec-simple-error"
    />
  </svg>
   
  <div
    class="ec-swift__content"
  >
    <div
      class="ec-swift__title"
    >
      Title example
    </div>
     
    <!---->
  </div>
   
  <!---->
</div>
`;

exports[`EcSwift should use the type "info" 1`] = `
<div
  class="ec-swift ec-swift--info ec-swift--is-responsive"
>
  <!---->
   
  <svg
    class="ec-swift__icon ec-icon"
  >
    <use
      xlink:href="#ec-simple-info"
    />
  </svg>
   
  <div
    class="ec-swift__content"
  >
    <div
      class="ec-swift__title"
    >
      Title example
    </div>
     
    <!---->
  </div>
   
  <!---->
</div>
`;

exports[`EcSwift should use the type "success" 1`] = `
<div
  class="ec-swift ec-swift--success ec-swift--is-responsive"
>
  <!---->
   
  <svg
    class="ec-swift__icon ec-icon"
  >
    <use
      xlink:href="#ec-simple-check"
    />
  </svg>
   
  <div
    class="ec-swift__content"
  >
    <div
      class="ec-swift__title"
    >
      Title example
    </div>
     
    <!---->
  </div>
   
  <!---->
</div>
`;

exports[`EcSwift should use the type "warning" 1`] = `
<div
  class="ec-swift ec-swift--warning ec-swift--is-responsive"
>
  <!---->
   
  <svg
    class="ec-swift__icon ec-icon"
  >
    <use
      xlink:href="#ec-simple-error"
    />
  </svg>
   
  <div
    class="ec-swift__content"
  >
    <div
      class="ec-swift__title"
    >
      Title example
    </div>
     
    <!---->
  </div>
   
  <!---->
</div>
`;
